File: sdks/node/src/plugins/kiroide-contract.ts
import os from 'node:os';
import path from 'node:path';
import {
  buildKiroIdeCommand,
  kiroIdeRuntimeReference,
  sameKiroIdeAgentId,
  sameKiroIdePath,
  type KiroIdeRuntimeReference,
} from './kiroide-command.js';
import { isObject, parseStrictJsonObject, type JsonObject } from './strict-json.js';

export const AGENT_KEY = 'kiroide';
export const AUDIT_SCRIPT = 'hook.js';
export const CONFIG_FILE = 'elydora-audit.json';
export const GUARD_SCRIPT = 'guard.js';
export const LEGACY_CONFIG_FILE = 'elydora-audit.kiro.hook';
export const HOOK_TIMEOUT_SECONDS = 10;

const GUARD_NAME = 'elydora-guard';
const AUDIT_NAME = 'elydora-audit';
const GUARD_DESCRIPTION = 'Block tool use when the Elydora agent is frozen';
const AUDIT_DESCRIPTION = 'Record tool use in the Elydora audit trail';
const KIRO_TRIGGERS = new Set([
  'SessionStart',
  'Stop',
  'PreToolUse',
  'PostToolUse',
  'PreTaskExec',
  'PostTaskExec',
  'UserPromptSubmit',
  'PostFileCreate',
  'PostFileSave',
  'PostFileDelete',
]);

export interface KiroIdeAction extends JsonObject {
  readonly type: 'command' | 'agent';
  readonly command?: string;
  readonly prompt?: string;
}

export interface KiroIdeHook extends JsonObject {
  readonly name: string;
  readonly description?: string;
  readonly trigger: string;
  readonly matcher?: string;
  readonly action: KiroIdeAction;
  readonly timeout?: number;
  readonly enabled?: boolean;
}

export interface KiroIdeDocument {
  readonly exists: boolean;
  readonly filePath: string;
  readonly root: JsonObject;
  readonly hooks: readonly KiroIdeHook[];
  readonly raw?: string;
}

export interface RenderedKiroIdeDocument {
  readonly document: KiroIdeDocument;
  readonly changed: boolean;
  readonly next?: string;
}

export interface KiroIdeRuntimeContract {
  readonly agentId: string;
  readonly guardPath: string;
  readonly auditPath: string;
}

export interface KiroIdePaths {
  readonly workspaceRoot: string;
  readonly kiroDirectory: string;
  readonly hooksDirectory: string;
  readonly configPath: string;
  readonly legacyConfigPath: string;
}

export function resolveKiroIdePaths(
  workspaceRoot = process.cwd(),
  homeDirectory = os.homedir(),
): KiroIdePaths {
  const resolvedWorkspace = path.resolve(workspaceRoot);
  const kiroDirectory = path.join(resolvedWorkspace, '.kiro');
  const hooksDirectory = path.join(kiroDirectory, 'hooks');
  return {
    workspaceRoot: resolvedWorkspace,
    kiroDirectory,
    hooksDirectory,
    configPath: path.join(hooksDirectory, CONFIG_FILE),
    legacyConfigPath: path.join(homeDirectory, '.kiro', 'hooks', LEGACY_CONFIG_FILE),
  };
}

function validateAction(value: unknown, label: string): KiroIdeAction {
  if (!isObject(value)) throw new Error(`${label} action must be an object`);
  if (value.type === 'command') {
    if (typeof value.command !== 'string' || !value.command) {
      throw new Error(`${label} command action requires a non-empty command`);
    }
    return value as KiroIdeAction;
  }
  if (value.type === 'agent') {
    if (typeof value.prompt !== 'string' || !value.prompt) {
      throw new Error(`${label} agent action requires a non-empty prompt`);
    }
    return value as KiroIdeAction;
  }
  throw new Error(`${label} action has unsupported type "${String(value.type)}"`);
}

function validateHook(value: unknown, index: number, documentLabel: string): KiroIdeHook {
  const label = `${documentLabel} hooks[${index}]`;
  if (!isObject(value)) throw new Error(`${label} must be an object`);
  if (typeof value.name !== 'string' || !value.name) throw new Error(`${label} requires a name`);
  if (typeof value.trigger !== 'string' || !KIRO_TRIGGERS.has(value.trigger)) {
    throw new Error(`${label} has unsupported trigger "${String(value.trigger)}"`);
  }
  if (value.description !== undefined && typeof value.description !== 'string') {
    throw new Error(`${label} description must be a string`);
  }
  if (value.matcher !== undefined && typeof value.matcher !== 'string') {
    throw new Error(`${label} matcher must be a string`);
  }
  if (value.timeout !== undefined
    && (!Number.isSafeInteger(value.timeout) || (value.timeout as number) < 0)) {
    throw new Error(`${label} timeout must be a non-negative integer`);
  }
  if (value.enabled !== undefined && typeof value.enabled !== 'boolean') {
    throw new Error(`${label} enabled must be a boolean`);
  }
  return { ...value, action: validateAction(value.action, label) } as KiroIdeHook;
}

export function parseKiroIdeDocument(
  filePath: string,
  raw: string,
  documentLabel = 'Kiro IDE hooks',
): KiroIdeDocument {
  const root = parseStrictJsonObject(raw, `${documentLabel} at ${filePath}`);
  if (root.version !== 'v1') throw new Error(`${documentLabel} version must be "v1": ${filePath}`);
  if (!Array.isArray(root.hooks)) throw new Error(`${documentLabel} field "hooks" must be an array: ${filePath}`);
  return {
    exists: true,
    filePath,
    root,
    hooks: root.hooks.map((hook, index) => validateHook(hook, index, documentLabel)),
    raw,
  };
}

export function createKiroIdeDocument(filePath: string): KiroIdeDocument {
  return { exists: false, filePath, root: {}, hooks: [] };
}

function exactKeys(value: JsonObject, keys: readonly string[]): boolean {
  return Object.keys(value).sort().join('|') === [...keys].sort().join('|');
}

function managedSpecification(name: string): {
  readonly description: string;
  readonly trigger: 'PreToolUse' | 'PostToolUse';
  readonly scriptName: string;
} | undefined {
  if (name === GUARD_NAME) {
    return { description: GUARD_DESCRIPTION, trigger: 'PreToolUse', scriptName: GUARD_SCRIPT };
  }
  if (name === AUDIT_NAME) {
    return { description: AUDIT_DESCRIPTION, trigger: 'PostToolUse', scriptName: AUDIT_SCRIPT };
  }
  return undefined;
}

function managedReference(hook: KiroIdeHook) {
  const specification = managedSpecification(hook.name);
  const owned = ownedReference(hook);
  if (!specification
    || !owned
    || !exactKeys(hook, ['name', 'description', 'trigger', 'matcher', 'action', 'timeout', 'enabled'])
    || hook.description !== specification.description
    || hook.trigger !== specification.trigger
    || hook.matcher !== '.*'
    || hook.timeout !== HOOK_TIMEOUT_SECONDS
    || hook.enabled !== true
    || !exactKeys(hook.action, ['type', 'command'])) return undefined;
  return owned;
}

function ownedReference(hook: KiroIdeHook) {
  const specification = managedSpecification(hook.name);
  if (!specification
    || hook.action.type !== 'command'
    || typeof hook.action.command !== 'string') return undefined;
  return kiroIdeRuntimeReference(hook.action.command, specification.scriptName);
}

export function requireAvailableKiroIdeHooks(
  hooks: readonly KiroIdeHook[],
  productLabel = 'Kiro IDE',
): void {
  for (const hook of hooks) {
    if (managedSpecification(hook.name) && !ownedReference(hook)) {
      throw new Error(`${productLabel} hook name "${hook.name}" conflicts with the Elydora contract`);
    }
  }
}

export function buildKiroIdeHook(
  name: typeof GUARD_NAME | typeof AUDIT_NAME,
  scriptPath: string,
): KiroIdeHook {
  const specification = managedSpecification(name)!;
  return {
    name,
    description: specification.description,
    trigger: specification.trigger,
    matcher: '.*',
    action: { type: 'command', command: buildKiroIdeCommand(scriptPath) },
    timeout: HOOK_TIMEOUT_SECONDS,
    enabled: true,
  };
}

export function withoutManagedKiroIdeHooks(
  hooks: readonly KiroIdeHook[],
  agentId?: string,
): KiroIdeHook[] {
  return hooks.filter((hook) => {
    const reference = ownedReference(hook);
    return !reference || (agentId !== undefined && !sameKiroIdeAgentId(reference.agentId, agentId));
  });
}

function entirelyManaged(document: KiroIdeDocument): boolean {
  return document.exists
    && exactKeys(document.root, ['version', 'hooks'])
    && document.hooks.length > 0
    && document.hooks.every((hook) => managedReference(hook) !== undefined);
}

export function renderKiroIdeDocument(
  document: KiroIdeDocument,
  hooks: readonly KiroIdeHook[],
): RenderedKiroIdeDocument {
  if (!document.exists && hooks.length === 0) return { document, changed: false };
  if (document.exists
    && hooks.length === document.hooks.length
    && hooks.every((hook, index) => hook === document.hooks[index])) {
    return { document, changed: false };
  }
  if (hooks.length === 0 && entirelyManaged(document)) return { document, changed: true };
  const next = `${JSON.stringify({ ...document.root, version: 'v1', hooks }, null, 2)}\n`;
  return { document, changed: next !== document.raw, next };
}

export function kiroIdeRuntimeContracts(
  hooks: readonly KiroIdeHook[],
): KiroIdeRuntimeContract[] {
  const guards = new Map<string, KiroIdeRuntimeReference[]>();
  const audits = new Map<string, KiroIdeRuntimeReference[]>();
  for (const hook of hooks) {
    if (!managedSpecification(hook.name)) continue;
    const reference = managedReference(hook);
    if (!reference) return [];
    const key = process.platform === 'win32' ? reference.agentId.toLowerCase() : reference.agentId;
    const target = hook.name === GUARD_NAME ? guards : audits;
    const entries = target.get(key) ?? [];
    entries.push(reference);
    target.set(key, entries);
  }
  const contracts: KiroIdeRuntimeContract[] = [];
  for (const [key, guard] of guards) {
    const audit = audits.get(key);
    if (guard.length !== 1 || audit?.length !== 1 || !guard[0] || !audit[0]) continue;
    contracts.push({
      agentId: guard[0].agentId,
      guardPath: guard[0].scriptPath,
      auditPath: audit[0].scriptPath,
    });
  }
  return contracts;
}

function legacyReference(command: unknown, scriptName: string) {
  if (typeof command !== 'string') return undefined;
  const match = /^node "([^"\r\n]+)"$/.exec(command);
  if (!match || !path.isAbsolute(match[1]) || path.basename(match[1]) !== scriptName) return undefined;
  const agentDirectory = path.dirname(match[1]);
  if (!sameKiroIdePath(path.dirname(agentDirectory), path.join(os.homedir(), '.elydora'))) {
    return undefined;
  }
  return { agentId: path.basename(agentDirectory), scriptPath: match[1] };
}

export function legacyKiroIdeRuntimeContract(raw: string, filePath: string): KiroIdeRuntimeContract | undefined {
  const root = parseStrictJsonObject(raw, `legacy Kiro IDE hook at ${filePath}`);
  if (!exactKeys(root, ['name', 'description', 'version', 'hooks'])
    || root.name !== 'Elydora Audit'
    || root.description !== 'Sends tool-use events to the Elydora tamper-evident audit platform'
    || root.version !== '1.0.0'
    || !isObject(root.hooks)
    || !exactKeys(root.hooks, ['pre_tool_use', 'post_tool_use'])) return undefined;
  const guardHook = root.hooks.pre_tool_use;
  const auditHook = root.hooks.post_tool_use;
  if (!isObject(guardHook)
    || !isObject(auditHook)
    || !exactKeys(guardHook, ['command', 'timeout_ms'])
    || !exactKeys(auditHook, ['command', 'timeout_ms'])
    || guardHook.timeout_ms !== 5000
    || auditHook.timeout_ms !== 5000) return undefined;
  const guard = legacyReference(guardHook.command, GUARD_SCRIPT);
  const audit = legacyReference(auditHook.command, AUDIT_SCRIPT);
  if (!guard || !audit || !sameKiroIdeAgentId(guard.agentId, audit.agentId)) return undefined;
  return { agentId: guard.agentId, guardPath: guard.scriptPath, auditPath: audit.scriptPath };
}

File: sdks/go/cmd/elydora/plugins/kiroide_contract.go
package plugins

import (
	"bytes"
	"encoding/json"
	"fmt"
	"math"
	"reflect"
)

const (
	kiroIdeAgentKey         = "kiroide"
	kiroIdeConfigFile       = "elydora-audit.json"
	kiroIdeLegacyFile       = "elydora-audit.kiro.hook"
	kiroIdeGuardScript      = "guard.js"
	kiroIdeAuditScript      = "hook.js"
	kiroIdeGuardName        = "elydora-guard"
	kiroIdeAuditName        = "elydora-audit"
	kiroIdeHookTimeout      = float64(10)
	kiroIdeGuardDescription = "Block tool use when the Elydora agent is frozen"
	kiroIdeAuditDescription = "Record tool use in the Elydora audit trail"
)

var kiroIdeTriggers = stringSet(
	"SessionStart",
	"Stop",
	"PreToolUse",
	"PostToolUse",
	"PreTaskExec",
	"PostTaskExec",
	"UserPromptSubmit",
	"PostFileCreate",
	"PostFileSave",
	"PostFileDelete",
)

type kiroIdeDocument struct {
	exists   bool
	filePath string
	root     map[string]any
	hooks    []map[string]any
	snapshot *managedFileSnapshot
}

type kiroIdeLegacyDocument struct {
	filePath string
	snapshot *managedFileSnapshot
	contract *managedRuntimeContract
}

type kiroIdeRenderedDocument struct {
	document *kiroIdeDocument
	changed  bool
	next     []byte
	remove   bool
}

type kiroIdeHookSpecification struct {
	description string
	trigger     string
	scriptName  string
}

func kiroIdeSpecification(name any) (kiroIdeHookSpecification, bool) {
	switch name {
	case kiroIdeGuardName:
		return kiroIdeHookSpecification{
			kiroIdeGuardDescription, "PreToolUse", kiroIdeGuardScript,
		}, true
	case kiroIdeAuditName:
		return kiroIdeHookSpecification{
			kiroIdeAuditDescription, "PostToolUse", kiroIdeAuditScript,
		}, true
	default:
		return kiroIdeHookSpecification{}, false
	}
}

func validateKiroIdeAction(value any, label string) error {
	action, ok := value.(map[string]any)
	if !ok || action == nil {
		return fmt.Errorf("%s action must be an object", label)
	}
	switch action["type"] {
	case "command":
		command, ok := action["command"].(string)
		if !ok || command == "" {
			return fmt.Errorf("%s command action requires a non-empty command", label)
		}
	case "agent":
		prompt, ok := action["prompt"].(string)
		if !ok || prompt == "" {
			return fmt.Errorf("%s agent action requires a non-empty prompt", label)
		}
	default:
		return fmt.Errorf(
			`%s action has unsupported type %q`,
			label,
			fmt.Sprint(action["type"]),
		)
	}
	return nil
}

func validateKiroIdeHook(value any, index int, label string) (map[string]any, error) {
	hook, ok := value.(map[string]any)
	if !ok || hook == nil {
		return nil, fmt.Errorf("%s hooks[%d] must be an object", label, index)
	}
	hookLabel := fmt.Sprintf("%s hooks[%d]", label, index)
	name, ok := hook["name"].(string)
	if !ok || name == "" {
		return nil, fmt.Errorf("%s requires a name", hookLabel)
	}
	trigger, ok := hook["trigger"].(string)
	if !ok {
		return nil, fmt.Errorf("%s requires a trigger", hookLabel)
	}
	if _, supported := kiroIdeTriggers[trigger]; !supported {
		return nil, fmt.Errorf("%s has unsupported trigger %q", hookLabel, trigger)
	}
	if value, exists := hook["description"]; exists {
		if _, ok := value.(string); !ok {
			return nil, fmt.Errorf("%s description must be a string", hookLabel)
		}
	}
	if value, exists := hook["matcher"]; exists {
		if _, ok := value.(string); !ok {
			return nil, fmt.Errorf("%s matcher must be a string", hookLabel)
		}
	}
	if value, exists := hook["timeout"]; exists {
		timeout, ok := value.(float64)
		if !ok || math.IsNaN(timeout) || math.IsInf(timeout, 0) ||
			timeout < 0 || timeout > 9007199254740991 || math.Trunc(timeout) != timeout {
			return nil, fmt.Errorf("%s timeout must be a non-negative integer", hookLabel)
		}
	}
	if value, exists := hook["enabled"]; exists {
		if _, ok := value.(bool); !ok {
			return nil, fmt.Errorf("%s enabled must be a boolean", hookLabel)
		}
	}
	if err := validateKiroIdeAction(hook["action"], hookLabel); err != nil {
		return nil, err
	}
	return hook, nil
}

func parseKiroIdeDocument(
	filePath string,
	snapshot *managedFileSnapshot,
) (*kiroIdeDocument, error) {
	if snapshot == nil {
		return &kiroIdeDocument{
			filePath: filePath,
			root:     map[string]any{},
			hooks:    []map[string]any{},
		}, nil
	}
	label := fmt.Sprintf("Kiro IDE hooks at %s", filePath)
	root, err := decodeStrictJSONObject(snapshot.contents, label)
	if err != nil {
		return nil, err
	}
	if root["version"] != "v1" {
		return nil, fmt.Errorf(`Kiro IDE hooks version must be "v1": %s`, filePath)
	}
	values, ok := root["hooks"].([]any)
	if !ok {
		return nil, fmt.Errorf(`Kiro IDE hooks field "hooks" must be an array: %s`, filePath)
	}
	hooks := make([]map[string]any, 0, len(values))
	for index, value := range values {
		hook, err := validateKiroIdeHook(value, index, "Kiro IDE hooks")
		if err != nil {
			return nil, err
		}
		hooks = append(hooks, hook)
	}
	return &kiroIdeDocument{
		exists: true, filePath: filePath, root: root, hooks: hooks,
		snapshot: snapshot,
	}, nil
}

func exactKiroIdeKeys(value map[string]any, keys ...string) bool {
	if len(value) != len(keys) {
		return false
	}
	for _, key := range keys {
		if _, exists := value[key]; !exists {
			return false
		}
	}
	return true
}

func ownedKiroIdeReference(
	hook map[string]any,
) (*managedScriptReference, error) {
	specification, managed := kiroIdeSpecification(hook["name"])
	if !managed {
		return nil, nil
	}
	action, ok := hook["action"].(map[string]any)
	if !ok || action["type"] != "command" {
		return nil, nil
	}
	command, ok := action["command"].(string)
	if !ok {
		return nil, nil
	}
	return kiroIdeRuntimeReferenceForCommand(command, specification.scriptName)
}

func managedKiroIdeReference(
	hook map[string]any,
) (*managedScriptReference, error) {
	specification, managed := kiroIdeSpecification(hook["name"])
	if !managed ||
		!exactKiroIdeKeys(
			hook,
			"name", "description", "trigger", "matcher", "action", "timeout", "enabled",
		) ||
		hook["description"] != specification.description ||
		hook["trigger"] != specification.trigger || hook["matcher"] != ".*" ||
		hook["timeout"] != kiroIdeHookTimeout || hook["enabled"] != true {
		return nil, nil
	}
	action, ok := hook["action"].(map[string]any)
	if !ok || !exactKiroIdeKeys(action, "type", "command") ||
		action["type"] != "command" {
		return nil, nil
	}
	return ownedKiroIdeReference(hook)
}

func requireAvailableKiroIdeHooks(hooks []map[string]any) error {
	for _, hook := range hooks {
		if _, managed := kiroIdeSpecification(hook["name"]); !managed {
			continue
		}
		reference, err := ownedKiroIdeReference(hook)
		if err != nil {
			return err
		}
		if reference == nil {
			return fmt.Errorf(
				"Kiro IDE hook name %q conflicts with the Elydora contract",
				hook["name"],
			)
		}
	}
	return nil
}

func buildKiroIdeHook(
	name string,
	runtimePath string,
	scriptPath string,
) (map[string]any, error) {
	specification, managed := kiroIdeSpecification(name)
	if !managed {
		return nil, fmt.Errorf("unsupported Elydora Kiro IDE hook name %q", name)
	}
	command, err := buildEncodedCommand("Kiro IDE", runtimePath, scriptPath)
	if err != nil {
		return nil, err
	}
	return map[string]any{
		"name": name, "description": specification.description,
		"trigger": specification.trigger, "matcher": ".*",
		"action":  map[string]any{"type": "command", "command": command},
		"timeout": kiroIdeHookTimeout, "enabled": true,
	}, nil
}

func withoutManagedKiroIdeHooks(
	hooks []map[string]any,
	agentID string,
) []map[string]any {
	filtered := make([]map[string]any, 0, len(hooks))
	for _, hook := range hooks {
		reference, err := ownedKiroIdeReference(hook)
		remove := err == nil && reference != nil &&
			(agentID == "" || sameManagedAgentID(reference.agentID, agentID))
		if !remove {
			filtered = append(filtered, hook)
		}
	}
	return filtered
}

func entirelyManagedKiroIdeDocument(document *kiroIdeDocument) bool {
	if document == nil || !document.exists ||
		!exactKiroIdeKeys(document.root, "version", "hooks") ||
		len(document.hooks) == 0 {
		return false
	}
	for _, hook := range document.hooks {
		reference, err := managedKiroIdeReference(hook)
		if err != nil || reference == nil {
			return false
		}
	}
	return true
}

func renderKiroIdeDocument(
	document *kiroIdeDocument,
	hooks []map[string]any,
) (*kiroIdeRenderedDocument, error) {
	if document == nil {
		return nil, fmt.Errorf("Kiro IDE rendering requires a hook document")
	}
	if !document.exists && len(hooks) == 0 {
		return &kiroIdeRenderedDocument{document: document}, nil
	}
	if reflect.DeepEqual(hooks, document.hooks) {
		return &kiroIdeRenderedDocument{document: document}, nil
	}
	if len(hooks) == 0 && entirelyManagedKiroIdeDocument(document) {
		return &kiroIdeRenderedDocument{
			document: document, changed: true, remove: true,
		}, nil
	}
	root := cloneJSONObject(document.root)
	root["version"] = "v1"
	values := make([]any, 0, len(hooks))
	for _, hook := range hooks {
		values = append(values, hook)
	}
	root["hooks"] = values
	next, err := json.MarshalIndent(root, "", "  ")
	if err != nil {
		return nil, fmt.Errorf("encode Kiro IDE hooks: %w", err)
	}
	next = append(next, '\n')
	validation := &managedFileSnapshot{contents: next}
	if _, err := parseKiroIdeDocument(document.filePath, validation); err != nil {
		return nil, fmt.Errorf("validate rendered Kiro IDE hooks: %w", err)
	}
	return &kiroIdeRenderedDocument{
		document: document,
		changed:  !document.exists || !bytes.Equal(next, document.snapshot.contents),
		next:     next,
	}, nil
}

func kiroIdeRuntimeContracts(
	hooks []map[string]any,
) ([]managedRuntimeContract, error) {
	managedHooks := make([]map[string]any, 0, 2)
	for _, hook := range hooks {
		_, managed := kiroIdeSpecification(hook["name"])
		if managed {
			managedHooks = append(managedHooks, hook)
		}
	}
	if len(managedHooks) != 2 {
		return []managedRuntimeContract{}, nil
	}
	nodePath, err := resolveNodeRuntime()
	if err != nil {
		return nil, err
	}
	var guard, audit *managedScriptReference
	for _, hook := range managedHooks {
		reference, err := currentKiroIdeReference(hook, nodePath)
		if err != nil {
			return nil, err
		}
		if reference == nil {
			return []managedRuntimeContract{}, nil
		}
		if hook["name"] == kiroIdeGuardName {
			if guard != nil {
				return []managedRuntimeContract{}, nil
			}
			guard = reference
		} else {
			if audit != nil {
				return []managedRuntimeContract{}, nil
			}
			audit = reference
		}
	}
	if guard == nil || audit == nil ||
		!sameManagedAgentID(guard.agentID, audit.agentID) {
		return []managedRuntimeContract{}, nil
	}
	return []managedRuntimeContract{{
		agentID: guard.agentID, guardPath: guard.scriptPath,
		auditPath: audit.scriptPath,
	}}, nil
}

func parseKiroIdeLegacyDocument(
	filePath string,
	snapshot *managedFileSnapshot,
) (*kiroIdeLegacyDocument, error) {
	document := &kiroIdeLegacyDocument{filePath: filePath, snapshot: snapshot}
	if snapshot == nil {
		return document, nil
	}
	label := fmt.Sprintf("legacy Kiro IDE hook at %s", filePath)
	root, err := decodeStrictJSONObject(snapshot.contents, label)
	if err != nil {
		return nil, err
	}
	if !exactKiroIdeKeys(root, "name", "hooks") || root["name"] != "Elydora Audit" {
		return document, nil
	}
	hooks, ok := root["hooks"].(map[string]any)
	if !ok || !exactKiroIdeKeys(hooks, "pre_tool_use", "post_tool_use") {
		return document, nil
	}
	guard, ok := hooks["pre_tool_use"].(map[string]any)
	if !ok || !exactKiroIdeKeys(guard, "command", "timeout_ms") ||
		guard["timeout_ms"] != float64(5000) {
		return document, nil
	}
	audit, ok := hooks["post_tool_use"].(map[string]any)
	if !ok || !exactKiroIdeKeys(audit, "command", "timeout_ms") ||
		audit["timeout_ms"] != float64(5000) {
		return document, nil
	}
	guardReference, err := legacyKiroIdeGoReference(guard["command"], kiroIdeGuardScript)
	if err != nil {
		return nil, err
	}
	auditReference, err := legacyKiroIdeGoReference(audit["command"], kiroIdeAuditScript)
	if err != nil {
		return nil, err
	}
	if guardReference == nil || auditReference == nil ||
		!sameManagedAgentID(guardReference.agentID, auditReference.agentID) {
		return document, nil
	}
	document.contract = &managedRuntimeContract{
		agentID: guardReference.agentID, guardPath: guardReference.scriptPath,
		auditPath: auditReference.scriptPath,
	}
	return document, nil
}
